### ISSUE TITLE
changing part of a declarative container restarts the whole container

### ISSUE THREAD

**Issue by ashkitten** [NixOS/nixpkgs#82217]
**Describe the bug**
changing parts of a declarative container should only restart the parts affected by the change

**To Reproduce**
1. create a declarative container with a systemd service
2. make a change to the service and `nixos-rebuild switch`
3. the entire container is stopped and restarted as part of the activation

**Expected behavior**
i expect the container to behave like the host system, restarting only the changed systemd unit

**Comment by flokli** [comment]
I think this was done on purpose in https://github.com/NixOS/nixpkgs/pull/62779 - see the explanation there.

I personally would also prefer a middle ground - if we detect we didn't change any of the host-specific config "around the container", just switch to the new generation in the container - but it's more complicated.

**Comment by flokli** [comment]
This is still an issue.

**Comment by flokli** [comment]
This is still an issue.

**Comment by flokli** [comment]
IIRC, this was mostly because of the networking bits, which are right now part of the start script.

https://github.com/NixOS/nixpkgs/issues/69414#issuecomment-909159831 from @Ma27 should actually fix this, too.

**Comment by Ma27** [comment]
Yes, it does :) I have a prototype where `switch-to-configuration.pl` actually checks what has changed in a nspawn-unit and decides whether a reload (i.e. a `switch-to-configuration` inside the container) is sufficient or a reboot is necessary (e.g. if bind-mounts have changed).

**Comment by noonien** [comment]
Any chance of sharing that script? Maybe some progress on this issue can be made. I have multiple containers running on the same machine, and every time a change is made to the common architecture, all of them are restarted, even though they should not be.

**Comment by Ma27** [comment]
That's a part of https://github.com/NixOS/nixpkgs/pull/140669.
